File: ts/candlestick-patterns/core/engine.ts
import { MembershipShape, muLess, muGreater, muGreaterEqual, muNear, muDirection } from '../../fuzzy/index.ts';
import { Criterion } from './criterion.ts';
import {
    DEFAULT_LONG_BODY, DEFAULT_VERY_LONG_BODY, DEFAULT_SHORT_BODY, DEFAULT_DOJI_BODY,
    DEFAULT_LONG_SHADOW, DEFAULT_VERY_LONG_SHADOW, DEFAULT_SHORT_SHADOW, DEFAULT_VERY_SHORT_SHADOW,
    DEFAULT_NEAR, DEFAULT_FAR, DEFAULT_EQUAL,
} from './defaults.ts';

// Minimum history size: 5-candle patterns + 10 default criterion period + 5 margin.
const MIN_HISTORY = 20;

/** OHLC bar. */
export interface OHLC {
    o: number;
    h: number;
    l: number;
    c: number;
}

/** Options for CandlestickPatterns engine. */
export interface CandlestickPatternsOptions {
    longBody?: Criterion;
    veryLongBody?: Criterion;
    shortBody?: Criterion;
    dojiBody?: Criterion;
    longShadow?: Criterion;
    veryLongShadow?: Criterion;
    shortShadow?: Criterion;
    veryShortShadow?: Criterion;
    near?: Criterion;
    far?: Criterion;
    equal?: Criterion;
    fuzzRatio?: number;
    shape?: MembershipShape;
}

/**
 * CriterionState maintains a running total for a single Criterion over a sliding window.
 */
export class CriterionState {
    readonly criterion: Criterion;
    private readonly ring: number[];
    private readonly ringSize: number;
    private ringStart: number = 0;
    private ringLen: number = 0;
    private total: number = 0;

    constructor(criterion: Criterion, maxShift: number) {
        this.criterion = criterion;
        this.ringSize = criterion.averagePeriod > 0 ? criterion.averagePeriod + maxShift : 0;
        this.ring = new Array(this.ringSize).fill(0);
    }

    push(o: number, h: number, l: number, c: number): void {
        if (this.ringSize === 0) return;
        const val = this.criterion.candleContribution(o, h, l, c);
        if (this.ringLen === this.ringSize) {
            this.total -= this.ring[this.ringStart];
            this.ring[this.ringStart] = val;
            this.ringStart = (this.ringStart + 1) % this.ringSize;
        } else {
            const idx = (this.ringStart + this.ringLen) % this.ringSize;
            this.ring[idx] = val;
            this.ringLen++;
        }
        this.total += val;
    }

    totalAt(shift: number): number {
        if (this.ringSize === 0 || this.criterion.averagePeriod <= 0) return 0.0;
        const period = this.criterion.averagePeriod;
        const n = this.ringLen;
        const end = n - shift;
        const start = end - period;
        if (start < 0 || end <= 0) return 0.0;
        let total = 0.0;
        for (let i = start; i < end; i++) {
            total += this.ring[(this.ringStart + i) % this.ringSize];
        }
        return total;
    }

    avg(shift: number, o: number, h: number, l: number, c: number): number {
        return this.criterion.averageValueFromTotal(this.totalAt(shift), o, h, l, c);
    }
}

/**
 * CandlestickPatternsEngine is the core engine with all internal state
 * and helper methods exposed for use by standalone pattern functions.
 *
 * This class is not intended for direct use by consumers — use the
 * CandlestickPatterns wrapper in the parent module instead.
 */
export class CandlestickPatternsEngine {
    readonly fuzzRatio: number;
    readonly shape: MembershipShape;

    readonly longBody: CriterionState;
    readonly veryLongBody: CriterionState;
    readonly shortBody: CriterionState;
    readonly dojiBody: CriterionState;
    readonly longShadow: CriterionState;
    readonly veryLongShadow: CriterionState;
    readonly shortShadow: CriterionState;
    readonly veryShortShadow: CriterionState;
    readonly near: CriterionState;
    readonly far: CriterionState;
    readonly equal: CriterionState;
    private readonly allStates: CriterionState[];

    private readonly history: OHLC[];
    private readonly histSize: number;
    private histStart: number = 0;
    private histLen: number = 0;
    count: number = 0;

    // Hikkake modified stateful fields.
    hikmodPatternResult: number = 0;
    hikmodPatternIdx: number = 0;
    hikmodConfirmed: boolean = false;
    hikmodLastSignal: number = 0;

    constructor(opts?: CandlestickPatternsOptions) {
        this.fuzzRatio = opts?.fuzzRatio ?? 0.2;
        this.shape = opts?.shape ?? MembershipShape.SIGMOID;

        this.longBody = new CriterionState(opts?.longBody?.copy() ?? DEFAULT_LONG_BODY.copy(), 5);
        this.veryLongBody = new CriterionState(opts?.veryLongBody?.copy() ?? DEFAULT_VERY_LONG_BODY.copy(), 5);
        this.shortBody = new CriterionState(opts?.shortBody?.copy() ?? DEFAULT_SHORT_BODY.copy(), 5);
        this.dojiBody = new CriterionState(opts?.dojiBody?.copy() ?? DEFAULT_DOJI_BODY.copy(), 5);
        this.longShadow = new CriterionState(opts?.longShadow?.copy() ?? DEFAULT_LONG_SHADOW.copy(), 5);
        this.veryLongShadow = new CriterionState(opts?.veryLongShadow?.copy() ?? DEFAULT_VERY_LONG_SHADOW.copy(), 5);
        this.shortShadow = new CriterionState(opts?.shortShadow?.copy() ?? DEFAULT_SHORT_SHADOW.copy(), 5);
        this.veryShortShadow = new CriterionState(opts?.veryShortShadow?.copy() ?? DEFAULT_VERY_SHORT_SHADOW.copy(), 5);
        this.near = new CriterionState(opts?.near?.copy() ?? DEFAULT_NEAR.copy(), 5);
        this.far = new CriterionState(opts?.far?.copy() ?? DEFAULT_FAR.copy(), 5);
        this.equal = new CriterionState(opts?.equal?.copy() ?? DEFAULT_EQUAL.copy(), 5);

        this.allStates = [
            this.longBody, this.veryLongBody, this.shortBody, this.dojiBody,
            this.longShadow, this.veryLongShadow, this.shortShadow, this.veryShortShadow,
            this.near, this.far, this.equal,
        ];

        let maxPeriod = 0;
        for (const s of this.allStates) {
            if (s.criterion.averagePeriod > maxPeriod) {
                maxPeriod = s.criterion.averagePeriod;
            }
        }
        let historySize = maxPeriod + 10;
        if (historySize < MIN_HISTORY) historySize = MIN_HISTORY;
        this.history = new Array(historySize);
        for (let i = 0; i < historySize; i++) {
            this.history[i] = { o: 0, h: 0, l: 0, c: 0 };
        }
        this.histSize = historySize;
    }

    /** Feeds a new OHLC bar into the engine (ring buffer + criterion states). */
    updateBar(o: number, h: number, l: number, c: number): void {
        const bar: OHLC = { o, h, l, c };
        if (this.histLen === this.histSize) {
            this.history[this.histStart] = bar;
            this.histStart = (this.histStart + 1) % this.histSize;
        } else {
            const idx = (this.histStart + this.histLen) % this.histSize;
            this.history[idx] = bar;
            this.histLen++;
        }
        for (const s of this.allStates) {
            s.push(o, h, l, c);
        }
        this.count++;
    }

    /** Gets OHLC of a bar. shift=1 is most recent, shift=2 is one before, etc. */
    bar(shift: number): OHLC {
        const idx = (this.histStart + this.histLen - shift) % this.histSize;
        return this.history[idx];
    }

    /** Checks if we have at least n bars in history. */
    has(n: number): boolean {
        return this.histLen >= n;
    }

    /** Checks if we have sufficient bars for a pattern requiring nCandles. */
    enough(nCandles: number, ...criteria: CriterionState[]): boolean {
        const avail = this.histLen - nCandles;
        for (const cs of criteria) {
            if (avail < cs.criterion.averagePeriod) return false;
        }
        return true;
    }

    /** Gets the criterion average value at a given shift. */
    avgCS(cs: CriterionState, shift: number): number {
        const b = this.bar(shift);
        return cs.avg(shift, b.o, b.h, b.l, b.c);
    }

    // Fuzzy membership helpers
    muLessCS(value: number, cs: CriterionState, shift: number): number {
        const avg = this.avgCS(cs, shift);
        let w = this.fuzzRatio * avg;
        if (avg <= 0.0) w = 0.0;
        return muLess(value, avg, w, this.shape);
    }

    muGreaterCS(value: number, cs: CriterionState, shift: number): number {
        const avg = this.avgCS(cs, shift);
        let w = this.fuzzRatio * avg;
        if (avg <= 0.0) w = 0.0;
        return muGreater(value, avg, w, this.shape);
    }

    muNearValue(value: number, target: number, cs: CriterionState, shift: number): number {
        const avg = this.avgCS(cs, shift);
        let w = this.fuzzRatio * avg;
        if (avg <= 0.0) w = 0.0;
        return muNear(value, target, w, this.shape);
    }

    muGeRaw(value: number, threshold: number, width: number): number {
        return muGreaterEqual(value, threshold, width, this.shape);
    }

    muGtRaw(value: number, threshold: number, width: number): number {
        return muGreater(value, threshold, width, this.shape);
    }

    muLtRaw(value: number, threshold: number, width: number): number {
        return muLess(value, threshold, width, this.shape);
    }

    muBullish(o: number, c: number, shift: number): number {
        const d = this.muDirectionRaw(o, c, shift);
        return d > 0.0 ? d : 0.0;
    }

    muBearish(o: number, c: number, shift: number): number {
        const d = this.muDirectionRaw(o, c, shift);
        return -d > 0.0 ? -d : 0.0;
    }

    muDirectionRaw(o: number, c: number, shift: number): number {
        const avg = this.avgCS(this.shortBody, shift);
        return muDirection(o, c, avg, 2.0);
    }

    /** Stateful update for hikkake_modified pattern. Called from Update(). */
    hikkakeModifiedUpdate(): void {
        if (this.count < 4) return;

        const b1 = this.bar(4);
        const b2 = this.bar(3);
        const b3 = this.bar(2);
        const b4 = this.bar(1);

        if (b2.h < b1.h && b2.l > b1.l &&
            b3.h < b2.h && b3.l > b2.l) {
            const nearAvg = this.avgCS(this.near, 3);
            if (b4.h < b3.h && b4.l < b3.l && b2.c <= b2.l + nearAvg) {
                this.hikmodPatternResult = 100.0;
                this.hikmodPatternIdx = this.count;
                return;
            }
            if (b4.h > b3.h && b4.l > b3.l && b2.c >= b2.h - nearAvg) {
                this.hikmodPatternResult = -100.0;
                this.hikmodPatternIdx = this.count;
                return;
            }
        }

        if (this.hikmodPatternResult !== 0 && this.count <= this.hikmodPatternIdx + 3) {
            const shift3rd = this.count - this.hikmodPatternIdx + 2;
            const b3rd = this.bar(shift3rd);

            if (this.hikmodPatternResult > 0 && b4.c > b3rd.h) {
                this.hikmodLastSignal = 200.0;
                this.hikmodPatternResult = 0.0;
                this.hikmodPatternIdx = 0;
                this.hikmodConfirmed = true;
                return;
            }
            if (this.hikmodPatternResult < 0 && b4.c < b3rd.l) {
                this.hikmodLastSignal = -200.0;
                this.hikmodPatternResult = 0.0;
                this.hikmodPatternIdx = 0;
                this.hikmodConfirmed = true;
                return;
            }
        }

        if (this.hikmodPatternResult !== 0 && this.count > this.hikmodPatternIdx + 3) {
            this.hikmodPatternResult = 0.0;
            this.hikmodPatternIdx = 0;
        }
    }
}

File: go/candlestickpatterns/core/engine.go
// Package core provides the CandlestickPatterns engine struct and all helper
// methods needed by the pattern implementations in the patterns/ subpackage.
package core

import (
	"zpano/fuzzy"
)

// Minimum history size: 5-candle patterns + 10 default criterion period + 5 margin.
const MinHistory = 20

// Options configures the CandlestickPatterns engine.
type Options struct {
	LongBody        *Criterion
	VeryLongBody    *Criterion
	ShortBody       *Criterion
	DojiBody        *Criterion
	LongShadow      *Criterion
	VeryLongShadow  *Criterion
	ShortShadow     *Criterion
	VeryShortShadow *Criterion
	Near            *Criterion
	Far             *Criterion
	Equal           *Criterion
	FuzzRatio       float64
	Shape           fuzzy.MembershipShape
}

// CandlestickPatterns is the candlestick pattern recognition engine.
//
// Provides streaming bar-by-bar evaluation of 61 Japanese candlestick patterns.
// Call Update(open, high, low, close) for each new bar, then call any pattern
// method to get the result for the current bar.
//
// Pattern methods return:
//
//	+100 for bullish match, -100 for bearish match, 0 for no match.
//	Some patterns return +50/-50 for unconfirmed signals (Hikkake, HikkakeModified).
type CandlestickPatterns struct {
	// Fuzzy configuration.
	FuzzRatio float64
	Shape     fuzzy.MembershipShape

	// Criteria states.
	LongBody        *CriterionState
	VeryLongBody    *CriterionState
	ShortBody       *CriterionState
	DojiBody        *CriterionState
	LongShadow      *CriterionState
	VeryLongShadow  *CriterionState
	ShortShadow     *CriterionState
	VeryShortShadow *CriterionState
	Near            *CriterionState
	Far             *CriterionState
	Equal           *CriterionState
	AllStates       []*CriterionState

	// Ring buffer of recent bars.
	History    []OHLC
	HistSize   int
	HistStart  int
	HistLen    int
	Count      int

	// Stateful pattern state: hikkake_modified
	HikmodPatternResult float64
	HikmodPatternIdx    int
	HikmodConfirmed     bool
	HikmodLastSignal    float64
}

func criterionOrDefault(c *Criterion, def Criterion) Criterion {
	if c != nil {
		return c.Copy()
	}
	return def.Copy()
}

// New creates a new CandlestickPatterns engine with the given options.
// Pass nil for default options.
func New(opts *Options) *CandlestickPatterns {
	var fuzzRatio float64 = 0.2
	var shape fuzzy.MembershipShape = fuzzy.Sigmoid

	var longBody, veryLongBody, shortBody, dojiBody Criterion
	var longShadow, veryLongShadow, shortShadow, veryShortShadow Criterion
	var nearC, farC, equalC Criterion

	if opts != nil {
		if opts.FuzzRatio != 0 {
			fuzzRatio = opts.FuzzRatio
		}
		shape = opts.Shape
		longBody = criterionOrDefault(opts.LongBody, DefaultLongBody)
		veryLongBody = criterionOrDefault(opts.VeryLongBody, DefaultVeryLongBody)
		shortBody = criterionOrDefault(opts.ShortBody, DefaultShortBody)
		dojiBody = criterionOrDefault(opts.DojiBody, DefaultDojiBody)
		longShadow = criterionOrDefault(opts.LongShadow, DefaultLongShadow)
		veryLongShadow = criterionOrDefault(opts.VeryLongShadow, DefaultVeryLongShadow)
		shortShadow = criterionOrDefault(opts.ShortShadow, DefaultShortShadow)
		veryShortShadow = criterionOrDefault(opts.VeryShortShadow, DefaultVeryShortShadow)
		nearC = criterionOrDefault(opts.Near, DefaultNear)
		farC = criterionOrDefault(opts.Far, DefaultFar)
		equalC = criterionOrDefault(opts.Equal, DefaultEqual)
	} else {
		longBody = DefaultLongBody.Copy()
		veryLongBody = DefaultVeryLongBody.Copy()
		shortBody = DefaultShortBody.Copy()
		dojiBody = DefaultDojiBody.Copy()
		longShadow = DefaultLongShadow.Copy()
		veryLongShadow = DefaultVeryLongShadow.Copy()
		shortShadow = DefaultShortShadow.Copy()
		veryShortShadow = DefaultVeryShortShadow.Copy()
		nearC = DefaultNear.Copy()
		farC = DefaultFar.Copy()
		equalC = DefaultEqual.Copy()
	}

	cp := &CandlestickPatterns{
		FuzzRatio:    fuzzRatio,
		Shape:        shape,
		LongBody:     NewCriterionState(longBody, 5),
		VeryLongBody: NewCriterionState(veryLongBody, 5),
		ShortBody:    NewCriterionState(shortBody, 5),
		DojiBody:     NewCriterionState(dojiBody, 5),
		LongShadow:   NewCriterionState(longShadow, 5),
		VeryLongShadow: NewCriterionState(veryLongShadow, 5),
		ShortShadow:    NewCriterionState(shortShadow, 5),
		VeryShortShadow: NewCriterionState(veryShortShadow, 5),
		Near:  NewCriterionState(nearC, 5),
		Far:   NewCriterionState(farC, 5),
		Equal: NewCriterionState(equalC, 5),
	}

	cp.AllStates = []*CriterionState{
		cp.LongBody, cp.VeryLongBody, cp.ShortBody, cp.DojiBody,
		cp.LongShadow, cp.VeryLongShadow, cp.ShortShadow, cp.VeryShortShadow,
		cp.Near, cp.Far, cp.Equal,
	}

	// History size: largest criterion period + 10, floored at MinHistory.
	maxPeriod := 0
	for _, s := range cp.AllStates {
		if s.Criterion.AveragePeriod > maxPeriod {
			maxPeriod = s.Criterion.AveragePeriod
		}
	}
	historySize := maxPeriod + 10
	if historySize < MinHistory {
		historySize = MinHistory
	}
	cp.History = make([]OHLC, historySize)
	cp.HistSize = historySize

	return cp
}

// UpdateBar feeds a new OHLC bar into the engine (ring buffer + criterion states).
// This does NOT call HikkakeModifiedUpdate — the caller must do that.
func (cp *CandlestickPatterns) UpdateBar(o, h, l, c float64) {
	bar := OHLC{o, h, l, c}
	if cp.HistLen == cp.HistSize {
		cp.History[cp.HistStart] = bar
		cp.HistStart = (cp.HistStart + 1) % cp.HistSize
	} else {
		idx := (cp.HistStart + cp.HistLen) % cp.HistSize
		cp.History[idx] = bar
		cp.HistLen++
	}
	for _, s := range cp.AllStates {
		s.Push(o, h, l, c)
	}
	cp.Count++
}

// Bar gets OHLC of a bar. shift=1 is most recent, shift=2 is one before, etc.
func (cp *CandlestickPatterns) Bar(shift int) OHLC {
	idx := (cp.HistStart + cp.HistLen - shift) % cp.HistSize
	return cp.History[idx]
}

// Has checks if we have at least n bars in history.
func (cp *CandlestickPatterns) Has(n int) bool {
	return cp.HistLen >= n
}

// Enough checks if we have sufficient bars for a pattern requiring nCandles
// plus the maximum average_period of the given criteria.
func (cp *CandlestickPatterns) Enough(nCandles int, criteria ...*CriterionState) bool {
	avail := cp.HistLen - nCandles
	for _, cs := range criteria {
		if avail < cs.Criterion.AveragePeriod {
			return false
		}
	}
	return true
}

// ---------------------------------------------------------------------------
// Criterion average helpers (shift is from the end, 1-based)
// ---------------------------------------------------------------------------

// AvgCS gets the criterion average value at a given shift from the most recent bar.
func (cp *CandlestickPatterns) AvgCS(cs *CriterionState, shift int) float64 {
	b := cp.Bar(shift)
	return cs.Avg(shift, b.O, b.H, b.L, b.C)
}

// ---------------------------------------------------------------------------
// Fuzzy membership helpers
// ---------------------------------------------------------------------------

// MuLess computes fuzzy 'value < avg' membership degree.
func (cp *CandlestickPatterns) MuLess(value float64, cs *CriterionState, shift int) float64 {
	avg := cp.AvgCS(cs, shift)
	w := cp.FuzzRatio * avg
	if avg <= 0.0 {
		w = 0.0
	}
	return fuzzy.MuLess(value, avg, w, cp.Shape)
}

// MuGreater computes fuzzy 'value > avg' membership degree.
func (cp *CandlestickPatterns) MuGreater(value float64, cs *CriterionState, shift int) float64 {
	avg := cp.AvgCS(cs, shift)
	w := cp.FuzzRatio * avg
	if avg <= 0.0 {
		w = 0.0
	}
	return fuzzy.MuGreater(value, avg, w, cp.Shape)
}

// MuNearValue computes fuzzy 'value ≈ target ± avg' membership degree.
func (cp *CandlestickPatterns) MuNearValue(value, target float64, cs *CriterionState, shift int) float64 {
	avg := cp.AvgCS(cs, shift)
	w := cp.FuzzRatio * avg
	if avg <= 0.0 {
		w = 0.0
	}
	return fuzzy.MuNear(value, target, w, cp.Shape)
}

// MuGeRaw computes fuzzy 'value >= threshold' with explicit width (no criterion).
func (cp *CandlestickPatterns) MuGeRaw(value, threshold, width float64) float64 {
	return fuzzy.MuGreaterEqual(value, threshold, width, cp.Shape)
}

// MuGtRaw computes fuzzy 'value > threshold' with explicit width (no criterion).
func (cp *CandlestickPatterns) MuGtRaw(value, threshold, width float64) float64 {
	return fuzzy.MuGreater(value, threshold, width, cp.Shape)
}

// MuLtRaw computes fuzzy 'value < threshold' with explicit width (no criterion).
func (cp *CandlestickPatterns) MuLtRaw(value, threshold, width float64) float64 {
	return fuzzy.MuLess(value, threshold, width, cp.Shape)
}

// MuBullish computes fuzzy degree of bullishness ∈ [0, 1].
func (cp *CandlestickPatterns) MuBullish(o, c float64, shift int) float64 {
	d := cp.MuDirectionRaw(o, c, shift)
	if d > 0.0 {
		return d
	}
	return 0.0
}

// MuBearish computes fuzzy degree of bearishness ∈ [0, 1].
func (cp *CandlestickPatterns) MuBearish(o, c float64, shift int) float64 {
	d := cp.MuDirectionRaw(o, c, shift)
	if -d > 0.0 {
		return -d
	}
	return 0.0
}

// MuDirectionRaw computes raw fuzzy direction ∈ [-1, +1].
func (cp *CandlestickPatterns) MuDirectionRaw(o, c float64, shift int) float64 {
	avg := cp.AvgCS(cp.ShortBody, shift)
	return fuzzy.MuDirection(o, c, avg, 2.0)
}

// HikkakeModifiedUpdate is called from Update() to track stateful hikkake_modified pattern.
func (cp *CandlestickPatterns) HikkakeModifiedUpdate() {
	if cp.Count < 4 {
		return
	}

	b1 := cp.Bar(4)
	b2 := cp.Bar(3)
	b3 := cp.Bar(2)
	b4 := cp.Bar(1)

	// Check for new pattern.
	if b2.H < b1.H && b2.L > b1.L &&
		b3.H < b2.H && b3.L > b2.L {
		nearAvg := cp.AvgCS(cp.Near, 3)
		// Bullish: 4th breaks low, 2nd close near its low.
		if b4.H < b3.H && b4.L < b3.L &&
			b2.C <= b2.L+nearAvg {
			cp.HikmodPatternResult = 100.0
			cp.HikmodPatternIdx = cp.Count
			return
		}
		// Bearish: 4th breaks high, 2nd close near its high.
		if b4.H > b3.H && b4.L > b3.L &&
			b2.C >= b2.H-nearAvg {
			cp.HikmodPatternResult = -100.0
			cp.HikmodPatternIdx = cp.Count
			return
		}
	}

	// No new pattern — check for confirmation.
	if cp.HikmodPatternResult != 0 &&
		cp.Count <= cp.HikmodPatternIdx+3 {
		shift3rd := cp.Count - cp.HikmodPatternIdx + 2
		b3rd := cp.Bar(shift3rd)

		if cp.HikmodPatternResult > 0 && b4.C > b3rd.H {
			cp.HikmodLastSignal = 200.0
			cp.HikmodPatternResult = 0.0
			cp.HikmodPatternIdx = 0
			cp.HikmodConfirmed = true
			return
		}
		if cp.HikmodPatternResult < 0 && b4.C < b3rd.L {
			cp.HikmodLastSignal = -200.0
			cp.HikmodPatternResult = 0.0
			cp.HikmodPatternIdx = 0
			cp.HikmodConfirmed = true
			return
		}
	}

	// If we passed the 3-bar window, reset.
	if cp.HikmodPatternResult != 0 &&
		cp.Count > cp.HikmodPatternIdx+3 {
		cp.HikmodPatternResult = 0.0
		cp.HikmodPatternIdx = 0
	}
}
